crates/net/network/src/fetch/client.rs

```diff
@@ -75,30 +78,46 @@ impl DownloadClient for FetchClient {
     }
 }
 
-#[async_trait::async_trait]
 impl HeadersClient for FetchClient {
+    type Output = HeadersFut;
+
     /// Sends a `GetBlockHeaders` request to an available peer.
-    async fn get_headers_with_priority(
+    fn get_headers_with_priority(
         &self,
         request: HeadersRequest,
         priority: Priority,
-    ) -> PeerRequestResult<BlockHeaders> {
+    ) -> Self::Output {
         let (response, rx) = oneshot::channel();
-        self.request_tx.send(DownloadRequest::GetBlockHeaders { request, response, priority })?;
-        rx.await?.map(WithPeerId::transform)
+        if self
+            .request_tx
+            .send(DownloadRequest::GetBlockHeaders { request, response, priority })
+            .is_ok()
+        {
+            Box::pin(ResolvedOneshotReceiver::from(rx).map(|r| r.map(WithPeerId::transform)))
+        } else {
+            Box::pin(future::ready(Err(RequestError::ChannelClosed)))
```

Comment:
this should work as well.
```suggestion
            Box::pin(futures::future::err(RequestError::ChannelClosed)))
```